pkg/simple/client/devops/fake/fakedevops.go

```diff
@@ -1,206 +1,290 @@
 package fake
 
 import (
+	"io/ioutil"
 	"kubesphere.io/kubesphere/pkg/simple/client/devops"
 	"net/http"
+	"net/url"
 	"strings"
 )
 
-type FakeDevops struct {
+type Devops struct {
 	Data map[string]interface{}
+
+	Projects map[string]interface{}
+}
+
+func New(projects ...string) *Devops {
+	d := &Devops{
+		Data:     nil,
+		Projects: map[string]interface{}{},
+	}
+	for _, p := range projects {
+		d.Projects[p] = true
+	}
+	return d
+}
+
+func (d *Devops) CreateDevOpsProject(projectId string) (string, error) {
+	if _, ok := d.Projects[projectId]; ok {
+		return projectId, nil
+	}
+	d.Projects[projectId] = true
+	return projectId, nil
+}
+
+func (d *Devops) DeleteDevOpsProject(projectId string) error {
+	if _, ok := d.Projects[projectId]; ok {
+		delete(d.Projects, projectId)
+		return nil
+	} else {
+		return &devops.ErrorResponse{
+			Body: []byte{},
+			Response: &http.Response{
+				Status:        "404 Not Found",
+				StatusCode:    404,
+				Proto:         "HTTP/1.1",
+				ProtoMajor:    1,
+				ProtoMinor:    1,
+				ContentLength: 50,
+				Header: http.Header{
+					"Foo": []string{"Bar"},
+				},
+				Body: ioutil.NopCloser(strings.NewReader("foo")), // shouldn't be used
+			},
+			Message: "",
+		}
+	}
+}
+
+func (d *Devops) GetDevOpsProject(projectId string) (string, error) {
+	if _, ok := d.Projects[projectId]; ok {
+		return projectId, nil
+	} else {
+		return "", &devops.ErrorResponse{
+			Body: []byte{},
+			Response: &http.Response{
+				Status:        "404 Not Found",
+				StatusCode:    404,
+				Proto:         "HTTP/1.1",
+				ProtoMajor:    1,
+				ProtoMinor:    1,
+				ContentLength: 50,
+				Header: http.Header{
+					"Foo": []string{"Bar"},
+				},
+				Body: ioutil.NopCloser(strings.NewReader("foo")), // shouldn't be used
+				Request: &http.Request{
+					Method: "",
+					URL: &url.URL{
+						Scheme:     "",
+						Opaque:     "",
+						User:       nil,
+						Host:       "",
+						Path:       "",
+						RawPath:    "",
+						ForceQuery: false,
+						RawQuery:   "",
+						Fragment:   "",
+					},
+				},
+			},
+			Message: "",
+		}
+	}
 }
 
-func NewFakeDevops(data map[string]interface{}) *FakeDevops {
-	var fakeData FakeDevops
+func NewFakeDevops(data map[string]interface{}) *Devops {
+	var fakeData Devops
 	fakeData.Data = data
 	return &fakeData
 }
 
 // Pipelinne operator interface
-func (d *FakeDevops) GetPipeline(projectName, pipelineName string, httpParameters *devops.HttpParameters) (*devops.Pipeline, error) {
+func (d *Devops) GetPipeline(projectName, pipelineName string, httpParameters *devops.HttpParameters) (*devops.Pipeline, error) {
```

Comment:
Why FakeDevops has all these no-op functions? How can we use this in unit test?

Reply:
When we create a devops project, we will create an admin namespace for the project, and the pipeline will be stored in the admin namespace.